### ISSUE TITLE
Configuration tip: run Docker container on Linux using host dbus and avahi-daemon

### ISSUE THREAD

**Issue by robcast** [mikebrady/shairport-sync#1387]
Not a bug report but a suggestion for an optional configuration for the current s6-based unstable-development Docker container to avoid starting additional instances of dbus and avahi-daemon in the container and use the host's dbus connection instead.

1. create an empty file `down`

2. create a `docker-compose.override.yml`:
```
version: "2.1"

services:
  shairport-sync:
    environment:
      - DBUS_SYSTEM_BUS_ADDRESS=unix:path=/run/dbus/system_bus_socket
    volumes:
      - /run/dbus:/run/dbus
      - ./down:/etc/services.d/01-dbus-daemon/down:ro
      - ./down:/etc/services.d/02-avahi/down:ro
```

This configuration inhibits the dbus and avahi daemons from starting and uses the hosts system dbus instead.

I don't know if this is useful for many people and if it should be added to the documentation somewhere. Maybe an advanced configuration tips section? I created this config to control the avahi-daemon on the host to suppress network-wide announcements for shairplay-sync on the internal Docker network interface ;-)

**Comment by mikebrady** [comment]
Thanks for this. It might take a little while to digest!

**Comment by charlesomer** [comment]
Hi, I think I've been able to replicate this but it looks like whilst it's possible to connect, it's not fully functional even though `/run/dbus/system_bus_socket` is shared between the host and the container. Did you have the same messages appear?

```
         0.019426354 "dbus-service.c:1057" *warning: could not acquire a Shairport Sync native D-Bus interface "org.gnome.ShairportSync.i789" on the system bus.
         0.001928958 "mpris-service.c:348" *warning: could not acquire an MPRIS interface named "org.mpris.MediaPlayer2.ShairportSync.i789" on the system bus.
```

**Comment by robcast** [comment]
@charlesomer I do see the same messages in my shairport-sync log but it works. I can connect to shairplay-sync in the network and I see "AirTunes Remote Audio" entries with `avahi-browse -art`.

I had some issues when I was playing around with this setup. The host avahi-daemon stopped to work and I had to reboot to get it to work again but after that it "works for me" :-)

**Comment by charlesomer** [comment]
Hmm, interesting. I think it’s best to leave this for now since I’m not sure how stable it would be etc.

**Comment by robcast** [comment]
I'll close this ticket as the exact solution doesn't work with the current config based on the new version of s6-overlay.

I have an updated config for my use. Re-open if you are interested ;-)

**Comment by dottedmag** [comment]
Thanks for the snippet. It still works just fine, with paths slightly changed:
- `/etc/s6-overlay/s6-rc.d/02-dbus/down`
- `/etc/s6-overlay/s6-rc.d/03-avahi/down`

**Comment by dottedmag** [comment]
I spoke too fast 😁 Will report back with an updated snippet that actually works once I get it working.

**Comment by dottedmag** [comment]
I've been able to make it work with the following configuration:
- Mount DBus socket directory as specified above
- Copy DBus policy configuration files (`shairport-sync*.conf`) to **host's** `/etc/dbus-1/system.d`, or shairport-sync won't be able acquire names on the system bus
- Suppress in-container DBus and Avahi.

I have done the last part by replacing entrypoint with the following shell script:
```
bail() {
  kill -TERM -1
}
trap bail TERM

(/usr/local/bin/nqptp; bail)&
(/usr/local/bin/shairport-sync "$@"; bail)&
wait
```

Some caveats:
- trap is needed to re-distribute SIGTERM on restart to all processs in the container - Docker sends signals only to the pid 1 of the container.
- `wait` is needed, one can't just run `shairport-sync` in foreground as a last line of the script: shell does not run trap handlers until the currently executing foreground command exits.
- running `bail` whenever nqptp or shairpotr-sync exits reliably takes down the container. I have played with `wait -n`, but it has unavoidable race conditions.
